packages/ast-spec/src/declaration/TSDeclareFunction/spec.ts

```diff
@@ -1,11 +1,35 @@
 import type { AST_NODE_TYPES } from '../../ast-node-types';
 import type { FunctionBase } from '../../base/FunctionBase';
-import type { BlockStatement } from '../../statement/BlockStatement/spec';
 
-// TODO(#1852) - async + declare are semantically invalid together
-export interface TSDeclareFunction extends FunctionBase {
+interface TSDeclareFunctionBase extends FunctionBase {
   type: AST_NODE_TYPES.TSDeclareFunction;
-  body: BlockStatement | undefined;
+  /**
+   * TS1183: An implementation cannot be declared in ambient contexts.
+   */
+  body: undefined;
+  /**
+   * Whether the declaration has `declare` modifier.
+   */
   declare: boolean;
   expression: false;
 }
+
+export interface TSDeclareFunctionWithDeclare extends TSDeclareFunctionBase {
+  /**
+   * TS1040: 'async' modifier cannot be used in an ambient context.
+   */
+  async: false;
+  declare: true;
+  /**
+   * TS1221: Generators are not allowed in an ambient context.
+   */
+  generator: false;
+}
+
+export interface TSDeclareFunctionNoDeclare extends TSDeclareFunctionBase {
```

Comment:
A comment to explain where this case is used would be good. Took me a minute to figure out where you could have this case.

```suggestion
/**
 * This case is used for overload declarations:
 * ```
 *    function foo(): void;
 * // ^^^^^^^^^^^^^^^^^^^^ TSDeclareFunction
 *    function foo() { }
 * ```
 */
export interface TSDeclareFunctionNoDeclare extends TSDeclareFunctionBase {
```